### DISCUSSION TITLE
如何使用 winsystray，其来源是哪里？

### DISCUSSION THREAD

**Discussion by lon91ong** [SeaHOH/GotoX#228] [问答/咨询] [answered]
想要参考源码修改自己的python程序, 总是遇到如题目的错误
挨个解压egg文件查看, 发现是`helpers-win32.egg`中
但是我的python环境中用pip安装`helpers`后, 依旧报错

请教这个`helpers-win32.egg`是哪里更新呢?

另外, [win_tray.py](launcher/win_tray.py)文件中使用的` SysTrayIcon`函数使用手册哪里找呢?

**Comment by lon91ong** [comment]
我使用循环生成两层菜单:
```
labs={'Sort1':[{'Name':'Jack', 'Args':'8866'},{'Name':'Bob', 'Args':'9900'}],'Sort2':[{'Name':'Jams', 'Args':'8686'},{'Name':'John', 'Args':'9090'}]}
main_menu = []
for k in labs.keys():
    main_menu.append((k, 'pass', MFS_DISABLED))
    for j in labs[k]:
        main_menu.append(('   '+j['Name'],lambda x: subOpen(SomeWeb_app+' '+j['Args'])))
    main_menu.append((None, '-'))
```
调用时接收到的`Args`参数始终是字典的最后一个`9090`
求大侠不吝赐教

**Reply by SeaHOH** [reply]
不要在闭包中使用会变化的变量，要使用固定的。如：
```py
    for j in labs[k]:
        main_menu.append(('   '+j['Name'],lambda x, args=j['Args']: subOpen(SomeWeb_app+' '+args)))
```

**Reply by lon91ong** [reply]
修改测试了一番, 问题还是没有解决, `SomeWeb_app`接收到的参数仍旧是`9090`, 无论点击哪个

**Reply by SeaHOH** [reply]
闭包调用太容易犯错，现在这个方法是对的。

**Reply by lon91ong** [reply]
能否改进一下[traybar.py](https://github.com/SeaHOH/GotoX/tree/master/python/site-packages/winsystray/traybar.py), 能够让响应函数接收参数, 或者 响应函数接收的`systray`参数能包含被点击的菜单项信息

**Reply by SeaHOH** [reply]
就是我现在改给你的这样做。

**Reply by lon91ong** [reply]
> 不要在闭包中使用会变化的变量，要使用固定的。如：
> 
> ```python
>     for j in labs[k]:
>         main_menu.append(('   '+j['Name'],lambda x, args=j['Args']: subOpen(SomeWeb_app+' '+args)))
> ```

这个确实是对的, 万分感谢!

**Comment by SeaHOH** [comment, marked as answer]
以下为 GotoX 中 [winsystray](https://github.com/SeaHOH/GotoX/tree/master/python/site-packages/winsystray) 的溯源路径：

- https://github.com/XX-net/XX-Net/tree/3.0.0/code/default/python27/1.0/lib/win32/systray

- https://github.com/Infinidat/infi.systray

- https://github.com/mhammond/pywin32/tree/main/win32/Demos
  参见 https://github.com/Infinidat/infi.systray#credit

winsystray 有添加修改以满足 GotoX 的使用需求，但并未单独发布到 PyPI，你可以将其包含到程序源码中。或者你也可以使用 infi.systray 作为依赖。

> 另外, [win_tray.py](https://github.com/SeaHOH/GotoX/blob/master/launcher/win_tray.py)文件中使用的` SysTrayIcon`函数使用手册哪里找呢?

请使用 `help()`，或参考 [traybar.py](https://github.com/SeaHOH/GotoX/tree/master/python/site-packages/winsystray/traybar.py)。
